mafContrib/microlensingMetric.py

```diff
@@ -0,0 +1,179 @@
+        self.mjdCol = mjdCol
+        self.m5Col = m5Col
+        self.filterCol = filterCol
+        self.nightCol = nightCol
+        self.ptsNeeded = ptsNeeded
+        self.detect_sigma = detect_sigma
+        # For now, let's just have a flat SED
+        # XXX--update to a stellar type
+        filters = 'ugrizy'
+        self.mags = {}
+        for filtername in filters:
+            self.mags[filtername] = rmag
+
+        cols = [self.mjdCol, self.m5Col, self.filterCol, self.nightCol]
+        super(MicrolensingMetric, self).__init__(col=cols,
+                                                 units='Detected, 0 or 1',
+                                                 metricName=metricName,
+                                                 **kwargs)
+
+    def run(self, dataSlice, slicePoint=None):
+        # Generate the lightcurve for this object
+        # make t a kind of simple way
+        t = dataSlice[self.mjdCol] - np.min(dataSlice[self.nightCol])
+        t = t - t.min()
+
+        amplitudes = microlensing_amplification(t, impact_parameter=slicePoint['impact_parameter'],
+                                                crossing_time=slicePoint['crossing_time'],
+                                                peak_time=slicePoint['peak_time'])
+
+        filters = np.unique(dataSlice[self.filterCol])
+        amplified_mags = amplitudes * 0
+
+        for filtername in filters:
+            infilt = np.where(dataSlice[self.filterCol] == filtername)[0]
+            amplified_mags[infilt] = self.mags[filtername] - 2.5*np.log10(amplitudes[infilt])
+
+        # The SNR of each point in the light curve
+        snr = m52snr(amplified_mags, dataSlice[self.m5Col])
+        # The magnitude uncertainties that go with amplified mags
+        mag_uncert = 2.5*np.log10(1+1./snr)
+
+        n_pre = []
+        for filtername in filters:
+            # observations pre-peak and in the given filter
+            infilt = np.where((dataSlice[self.filterCol] == filtername) & (t < slicePoint['peak_time']))[0]
+            # Broadcast to calc the mag_i - mag_j
+            diffs = amplified_mags[infilt] - amplified_mags[infilt][:, np.newaxis]
+            diffs_uncert = np.sqrt(mag_uncert[infilt]**2 + mag_uncert[infilt][:, np.newaxis]**2)
+
+            # Calculating this as a catalog-level detection. In theory,
+            # we could have a high SNR template image, so there would be
+            # little to no additional uncertianty from the subtraction.
+
+            sigma_above = np.abs(diffs)/diffs_uncert
+            # divide by 2 because array has i,j and j,i
+            n_above = np.size(np.where(sigma_above > self.detect_sigma)[0])/2
+            n_pre.append(n_above)
+
+        npts = np.sum(n_pre)
+        if npts >= self.ptsNeeded:
+            return 1
+        else:
+            return 0
+
+
+def microlensingSlicer(min_crossing_time=1, max_crossing_time=10, t_start=1,
```

Comment:
Related to comment above, can we maybe make a parent class for these? This time this looks like a function, but because the name matches our current set of XSlicer pattern, my brain is still expecting this to *be* the slicer, not returning the slicer. So could rename to 'setupMicrolensingSlicer' or something at least for now?